templated-this.md

```diff
@@ -1,151 +1,315 @@
-
-TBD: more here.
-
-## Impact on the Standard
-   
-TBD: A bunch of stuff in section 8.1.5 [expr.prim.lambda].
-
-## Acknowledgments
-
-Thanks to the following for their help and guidance:
-
-Louis Dionne
-
-## References
-
-We should probably look up the paper that added `*this` capture and why. And any
-other salient papers.
-
+<pre>
+Document number: PxxxxR0
+Date:            2017-10-03
+Project:         Programming Language C++, Evolution Working Group
+Reply-to:        Ben Deane <ben at elbeno dot com>              
+                 Gašper Ažman <gasper dot azman at gmail dot com>
+</pre>
+
+Template `this`! (Divining the value-category of `*this`)
+================
+
+Introduction
+------------
+
+We propose a new mechanism for specifying the value category of an instance of a
+class which is visible from inside a member function of that class.
+In other words, a way to tell from within a member function whether one's `this`
+points to an rvalue or an lvalue, and whether it is const or volatile.
+
+Motivation
+----------
+
+The existing mechanism for this (adding a reference, rvalue-reference, const, or
+volatile qualifier (_cv-ref qualifiers_) suffixes to a member function) suffers
+from problems.
+
+### It is verbose. 
+
+A common task in writing a class is providing a "getter function" to access a
+contained member. Authors that care about performance want to provide getter
+functions that take advantage of move semantics in the case where the class
+instance is an rvalue, and so end up writing several of these functions, which
+only differ in the cv-ref qualifiers .
+
+
+### There is no way to write a reference qualifier for a lambda expression.
+
+When writing a lambda expression, it is impossible to know whether it is safe to
+move from captured-by-copy members of the closure object. This impacts the
+performance of lambda expressions. One important scenario is when using lambda
+expressions as parts of a chain of asynchronous continuations that carry a value
+through a computation. Without knowing the value category of a closure object
+from inside its function call operator, the implementer has a dilemma: incur a
+copy, or use a move, and impose a requirement that is unenforceable in code and
+sets hidden traps for the unwary.
+
+
+Design Decisions
+----------------
+   
+In addition to solving the existing problems, desirable properties of a solution
+are:
+
+* It should work the same way for member functions and lambda expressions.
+
+* It should work like existing practice as much as possible, while adding as
+  little extra syntax as possible. Using `auto&& this` follows existing patterns
+  for dealing with forwarding references. Optionally adding "this" as the first
+  parameter fits with many programmers' mental model of the `this` pointer being
+  the first parameter to member functions "under the hood" and is comparable to
+  usage in other languages, e.g. Python.
```

Comment:
This para introduces our proposed solution, which I think is why I put the proposed solution first. I'm fine with putting design decisions first but maybe we should edit this point to omit the actual solution proposed, and expand on it later.

Reply:
The guidance on writing papers was "introduce problem first, and all the requirements" and then make a big line in your paper and then introduce a solution.

Reply:
Agree. I think we should say basically just the first line of this bullet here.

Reply:
Put the rest into a separate section "Teachability Implications"